src/device_imxrt.c

```diff
@@ -0,0 +1,278 @@
+#ifndef FLEXSPI_LUT_OPERAND0
+#define FLEXSPI_LUT_OPERAND0(x) (((uint32_t) (((uint32_t) (x)))) & 0xFFU)
+#define FLEXSPI_LUT_NUM_PADS0(x) \
+  (((uint32_t) (((uint32_t) (x)) << 8U)) & 0x300U)
+#define FLEXSPI_LUT_OPCODE0(x) \
+  (((uint32_t) (((uint32_t) (x)) << 10U)) & 0xFC00U)
+#define FLEXSPI_LUT_OPERAND1(x) \
+  (((uint32_t) (((uint32_t) (x)) << 16U)) & 0xFF0000U)
+#define FLEXSPI_LUT_NUM_PADS1(x) \
+  (((uint32_t) (((uint32_t) (x)) << 24U)) & 0x3000000U)
+#define FLEXSPI_LUT_OPCODE1(x) \
+  (((uint32_t) (((uint32_t) (x)) << 26U)) & 0xFC000000U)
+#endif
+
+#define FLEXSPI_NOR_INSTANCE 0
+
+#if MG_DEVICE == MG_DEVICE_RT1020
+typedef struct {
+  uint32_t version;
+  int (*init)(uint32_t instance, flexspi_nor_config_t *config);
+  int (*program)(uint32_t instance, flexspi_nor_config_t *config,
+                      uint32_t dst_addr, const uint32_t *src);
+  uint32_t reserved;
+  int (*erase)(uint32_t instance, flexspi_nor_config_t *config,
+                    uint32_t start, uint32_t lengthInBytes);
+  uint32_t reserved2;
+  int (*update_lut)(uint32_t instance, uint32_t seqIndex,
+                         const uint32_t *lutBase, uint32_t seqNumber);
+  int (*xfer)(uint32_t instance, char *xfer);
+  void (*clear_cache)(uint32_t instance);
+} flexspi_nor_driver_interface_t;
+#elif MG_DEVICE == MG_DEVICE_RT1060
+typedef struct {
+  uint32_t version;
+  int (*init)(uint32_t instance, flexspi_nor_config_t *config);
+  int (*program)(uint32_t instance, flexspi_nor_config_t *config,
+                      uint32_t dst_addr, const uint32_t *src);
+  int (*erase_all)(uint32_t instance, flexspi_nor_config_t *config);
+  int (*erase)(uint32_t instance, flexspi_nor_config_t *config,
+                    uint32_t start, uint32_t lengthInBytes);
+  int (*read)(uint32_t instance, flexspi_nor_config_t *config,
+                   uint32_t *dst, uint32_t addr, uint32_t lengthInBytes);
+  void (*clear_cache)(uint32_t instance);
+  int (*xfer)(uint32_t instance, char *xfer);
+  int (*update_lut)(uint32_t instance, uint32_t seqIndex,
+                         const uint32_t *lutBase, uint32_t seqNumber);
+  int (*get_config)(uint32_t instance, flexspi_nor_config_t *config,
+                         uint32_t *option);
+} flexspi_nor_driver_interface_t;
+#endif
+
+typedef struct {
+  const uint32_t version;  // Bootloader version number
+  const char *copyright;   // Bootloader Copyright
+  void (*runBootloader)(
+      void *arg);             // Function to start the bootloader executing
+  const uint32_t *reserved0;  // Reserved
+  const flexspi_nor_driver_interface_t
+      *flexSpiNorDriver;         // FlexSPI NOR Flash API
+  const uint32_t *reserved1[2];  // Reserved
+} bootloader_api_entry_t;
+
+#define bootloader (*(bootloader_api_entry_t **) (0x0020001c))
+#define flexspi_nor bootloader->flexSpiNorDriver
+
+
+static bool s_flash_irq_disabled;
+
+MG_IRAM void *mg_flash_start(void) {
+  return (void *) 0x0;
+}
```

Comment:
What if mg_flash_start() returns 0x60000000 ?

Reply:
Yes, you are right. Initially, there was the problem that the flash is offset by this value (0x60000000). So, there appeared two cases: when you wanted to write (using the ROM API), you had to translate this address by subtracting 0x60000000 from it. The other case was when you wanted to read and it this case you just accessed it at 0x60000000 (no subtraction required).

What I wanted to do by using that macro was to avoid over-complicating myself in the mg_flash_write / mg_flash_erase functions with having to worry about translating the write address when using the ROM API, so I added that macro which worked for any case (for flashes that require no translation, like ST, it's value is 0 and changes nothing)

Anyway, I realise now that it can work even without adding the macro. We'll stick to the original mg_flash_* API like Sergey commented above. I'll make the address translations (for ROM API calls to write/erase) in the mg_flash_write/erase functions and leave the rest of the code untouched.